django/contrib/postgres/operations.py

```diff
@@ -75,3 +79,58 @@ class UnaccentExtension(CreateExtension):
 
     def __init__(self):
         self.name = 'unaccent'
+
+
+class EnsureNotInTransactionMixin:
+    def _ensure_not_in_transaction(self, schema_editor):
+        if schema_editor.connection.in_atomic_block:
+            raise NotSupportedError(
+                'The %s operation cannot be executed inside a transaction '
+                '(set atomic = False on the migration).' % self.__class__.__name__
+            )
+
+
+class AddIndexConcurrently(EnsureNotInTransactionMixin, AddIndex):
+    """Create an index using Postgres's CREATE INDEX CONCURRENTLY syntax."""
+
+    def describe(self):
+        return re.sub(r'^Create', 'Concurrently create', super().describe())
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        self._ensure_not_in_transaction(schema_editor)
+
+        model = to_state.apps.get_model(app_label, self.model_name)
+        sql = re.sub('^CREATE INDEX', 'CREATE INDEX CONCURRENTLY', str(self.index.create_sql(model, schema_editor)))
+        schema_editor.execute(sql)
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        self._ensure_not_in_transaction(schema_editor)
+
+        model = from_state.apps.get_model(app_label, self.model_name)
+        sql = re.sub('^DROP INDEX', 'DROP INDEX CONCURRENTLY', str(self.index.remove_sql(model, schema_editor)))
+        schema_editor.execute(sql)
```

Comment:
Not a big fan of the regex replacing here. I guess the alternative could be to:

- Changed the PostgreSQL backend `add_index` and `create_index` method to accept a `concurrently` kwarg.
- Change `database_forwards` and `database_backwards` methods to pass `concurrently=True`.
- Change `Index.create_sql` and `remove_index` to accept `**kwargs` and pass them along to `_create_index_sql` and `_delete_index_sql`
- Change the PostgreSQL backend `_create_index_sql` method to accept a `concurrently` kwarg and pass `sql=self.sql_create_index_concurrently` to super when provided.
- Adapt the PostgreSQL and base backend `_delete_index_sql` method to be able to perform the same `sql` override based on `concurrently`.

Thoughts?

Reply:
I think I followed your advice here, to some extend.